lib/src/presentation/tcp_scanner_task.dart

```diff
@@ -0,0 +1,106 @@
+import 'dart:math';
+
+import 'tcp_scanner_task_report.dart';
+import '../domain/entities/report.dart';
+import '../domain/interactor/scanner_isolate_interactor.dart';
+import '../presentation/tcp_scanner_task_exception.dart';
+
+class TcpScannerTask {
+  final String host;
+  late final List<int> ports;
+  final Duration socketTimeout;
+  final bool shuffle;
+  late final int parallelism;
+  bool _isRunning = false;
+  final List<ScannerIsolateInteractor> _scanners = [];
+
+  bool get isRunning => _isRunning;
+
+  TcpScannerTask(this.host, List<int> ports,
+      {this.socketTimeout = const Duration(seconds: 1), this.shuffle = false, int parallelism = 4}) {
+    //  Copy ports list and shuffle them if it's needed
+    var portsList = ports.toSet().toList();
+    if (shuffle) portsList.shuffle();
+    this.ports = portsList;
+    // Calculate number of isolates. The number of isolates can't be more than the number of ports.
+    this.parallelism = min(parallelism, ports.length);
+  }
+
+  /// Start scanning task
+  Future<TcpScannerTaskReport> start() async {
+    if (_isRunning) {
+      throw TcpScannerTaskException('Scanning is in progress');
+    }
+    _isRunning = true;
+    // Split the list of ports into sublists
+    var portsPerScanner = (ports.length / parallelism).ceil();
```

Comment:
@shpak86 

I got this issue when running your code

```
[VERBOSE-2:ui_dart_state.cc(199)] Unhandled Exception: Unsupported operation: Infinity or NaN toInt
#0      double.toInt (dart:core-patch/double.dart:219:36)
#1      double.ceil (dart:core-patch/double.dart:184:32)
#2      TcpScannerTask.start (package:tcp_scanner/src/presentation/tcp_scanner_task.dart:36:56)
```

Reply:
The problem probably is in the empty `ports` list or `parallelism` parameter was 0. Please check these arguments. I added exception throwing to the constructor so you'll be able to get same additional information in case of error.
Try to cover your code in try-catch to catch constructor exceptions and don't forget `catchError` to catch exceptions threw by scan()
```dart 
  try {
    await TcpScannerTask(host, ports, shuffle: true, parallelism: 2)
        .start()
        .then((){})
        // Catch errors during the scan
        .catchError((error) => stderr.writeln(error));
  } on TcpScannerTaskException catch (e) {
    // Here you can catch exceptions threw in the constructor
    stderr.writeln('Error: ${e.cause}');
  }
```

Reply:
@shpak86 
It's resolved. Thank you so much!